cinn/runtime/buffer.h

```diff
@@ -73,10 +73,7 @@ class Buffer {
     CHECK(data_) << "alloc buffer failed";
   }
   //! Deallocate the memory in host device.
-  void DeallocHost() {
-    if (data_) delete data_;
-    data_ = nullptr;
-  }
+  void DeallocHost() { data_ = nullptr; }
```

Comment:
delete data_;
data_ = nullptr;

Reply:
:thinking:
* How do you think about to [work also with smart pointers](https://en.cppreference.com/w/cpp/memory/unique_ptr/reset "Documentation")?
* Can a statement like “`data_.reset(nullptr);`” be used instead?

Reply:
Hi, @elfring

Since it is raw pointer here, we shouldn't set it to `nullptr` without `delete data_`. Otherwise it would be memory leak. Please change as @luotao1 commented.

As for the smart pointer suggestion, CINN works at low level to handle the buffers by itself, so we didn't use smart pointer in buffer. Thank you for your suggestion, but we won't change it to smart pointer currently.

Reply:
>Thank you for your suggestion, but we won't change it to smart pointer currently.

:thinking: Under which circumstances would you [take further applications into account also for smart pointers](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r11-avoid-calling-new-and-delete-explicitly "C++ guideline “R.11: Avoid calling new and delete explicitly”")?

Reply:
@enkilee 

delete data_;
data_ = nullptr;

reason: https://github.com/PaddlePaddle/CINN/pull/1140#discussion_r1064257166